File: JavaScript/1049-Last Stone Weight II.js
/**
 * @param {number[]} stones
 * @return {number}
 */
var lastStoneWeightII = function(stones) {

//knapsack
    
    const summ = stones.reduce((acc, x) => acc+x);
    const upperbound = Math.floor(summ/2);
    let dp = new Array(upperbound+1).fill(false);
    dp[0] = true;
        
    for (let stone of stones){
        for (let i=dp.length-1; i>0; i--){
            if (i-stone < 0) break;
            if (dp[i-stone]) dp[i] = true;
        }
        if (dp[dp.length-1]) return summ%2;
    }
    
    for (let i=dp.length-1; i>=0; i--){
        if (dp[i]) return summ-2*i;
    }
};

File: 1049-Last Stone Weight II.py
class Solution(object):
    def lastStoneWeightII(self, stones):
        """
        :type stones: List[int]
        :rtype: int
        """
# knapsack
        summ = sum(stones)
        upperbound = summ//2
        dp = [False]*(upperbound+1)
        dp[0] = True
        
        for stone in stones:
            for i in range(len(dp)-1, 0, -1):
                if i-stone < 0:
                    break
                if dp[i-stone]:
                    dp[i] = True
            if dp[-1]:
                return summ%2
        
        for i in range(len(dp)-1, -1, -1):
            if dp[i]:
                return summ-2*i